api/agent/agent.go

```diff
@@ -266,88 +266,114 @@ func (a *agent) getSlot(ctx context.Context, call *call) (Slot, error) {
 
 	isHot := protocol.IsStreamable(protocol.Protocol(call.Format))
 	if isHot {
-		// For hot requests, we use a long lived slot queue, which we use to manage hot containers
-		call.slots = a.slotMgr.getHotSlotQueue(call)
 		start := time.Now()
 
-		call.slots.enterState(SlotQueueWaiter)
-		s, err := a.launchHot(ctx, call)
-		call.slots.exitStateWithLatency(SlotQueueWaiter, uint64(time.Now().Sub(start).Seconds()*1000))
+		// For hot requests, we use a long lived slot queue, which we use to manage hot containers
+		var isNew bool
+		call.slots, isNew = a.slotMgr.getSlotQueue(call)
+		if isNew {
+			go a.hotLauncher(ctx, call)
+		}
 
+		s, err := a.waitHot(ctx, call)
+		call.slots.exitStateWithLatency(SlotQueueWaiter, uint64(time.Now().Sub(start).Seconds()*1000))
 		return s, err
 	}
 
 	return a.launchCold(ctx, call)
 }
 
-// launchHot checks with slot queue to see if a new container needs to be launched and waits
-// for available slots in the queue for hot request execution.
-func (a *agent) launchHot(ctx context.Context, call *call) (Slot, error) {
+// hotLauncher is spawned in a go routine for each slot queue to monitor stats and launch hot
+// containers if needed. Upon shutdown or activity timeout, hotLauncher exits and during exit,
+// it destroys the slot queue.
+func (a *agent) hotLauncher(ctx context.Context, call *call) {
+
+	// Let use 60 minutes or 2 * IdleTimeout as hot queue idle timeout, pick
+	// whichever is longer. If in this time, there's no activity, then
+	// we destroy the hot queue.
+	timeout := time.Duration(60) * time.Minute
+	idleTimeout := time.Duration(call.IdleTimeout) * time.Second * 2
+	if timeout < idleTimeout {
+		timeout = idleTimeout
+	}
 
+	logger := common.Logger(ctx)
+	logger.WithField("launcher_timeout", timeout).Info("Hot function launcher starting")
 	isAsync := call.Type == models.TypeAsync
 
-launchLoop:
+	defer call.slots.destroySlotQueue()
+
 	for {
-		// Check/evaluate if we need to launch a new hot container
-		doLaunch, stats := call.slots.isNewContainerNeeded()
-		common.Logger(ctx).WithField("stats", stats).Debug("checking hot container launch ", doLaunch)
+		select {
+		case <-a.shutdown: // server shutdown
+			return
+		case <-time.After(timeout):
+			if a.slotMgr.deleteSlotQueue(call.slots) {
+				logger.Info("Hot function launcher timed out")
+				return
+			}
+		case <-call.slots.signaller:
+		}
 
-		if doLaunch {
-			ctxToken, tokenCancel := context.WithCancel(context.Background())
+		isNeeded, stats := call.slots.isNewContainerNeeded()
+		logger.WithField("stats", stats).Debug("Hot function launcher stats")
+		if !isNeeded {
+			continue
+		}
 
-			// wait on token/slot/timeout whichever comes first
-			select {
-			case tok, isOpen := <-a.resources.GetResourceToken(ctxToken, call.Memory, isAsync):
-				tokenCancel()
-				if !isOpen {
-					return nil, models.ErrCallTimeoutServerBusy
-				}
+		resourceCtx, cancel := context.WithCancel(context.Background())
```

Comment:
could remove this and just use `a.resources.GetResourceToken(context.Background(), call.Memory, isAsync)`, which removes the 3 `cancel()` func calls in each select case below. 

it would be nice to keep some context's logger and span info, adding either of these won't require a `cancel` func either, though. this thread seems like it could add app / route / image to the logger and have its own span that is a `FollowsFrom` (terminology: https://github.com/opentracing/specification/blob/master/specification.md ) of the enclosing context's span.

Reply:
each hot container launch would also be a `FollowsFrom` span child of the span for this thread (would be nice to track these so we can watch causal chain of incoming calls vs container launches)

Reply:
Could you explain why we don't need cancel() in GetResourceToken() context? For two cases, timeout and shutdown, don't we need to cancel that go routine in GetResourceToken?

Reply:
d'oh, that's right.